gnucash/gnome-utils/gnc-gtk-utils.hpp

```
@@ -0,0 +1,130 @@
+};
+
+// Custom iterator class
+template <typename ModelType>
+class GtkTreeModelIterator
+{
+public:
+    GtkTreeModelIterator(GtkTreeModel* model, std::optional<GtkTreeIter> iter) : m_model(model), m_iter(iter) {}
+
+    GtkTreeModelIterator& operator++()
+    {
+        if (!m_iter.has_value())
+            throw "no value, cannot increment";
+        if (!gtk_tree_model_iter_next(m_model, &m_iter.value()))
+            m_iter = std::nullopt;
+        return *this;
+    }
+
+    ModelType operator*() const
+    {
+        if (!m_iter.has_value())
+            throw "no value, cannot dereference";
+        return ModelType (m_model, *m_iter);
+    }
+
+    bool has_value() { return m_iter.has_value(); };
+
+    bool operator==(const GtkTreeModelIterator& other) const
+    {
+        if (m_model != other.m_model)
+            return false;
+        if (!m_iter.has_value() && !other.m_iter.has_value())
+            return true;
+        if (!m_iter.has_value() || !other.m_iter.has_value())
+            return false;
+        return (m_iter->stamp == other.m_iter->stamp) &&
+            (m_iter->user_data == other.m_iter->user_data) &&
+            (m_iter->user_data2 == other.m_iter->user_data2) &&
+            (m_iter->user_data3 == other.m_iter->user_data3);
+    }
+
+    bool operator!=(const GtkTreeModelIterator& other) const { return !(*this == other); }
+
+private:
+    GtkTreeModel* m_model;
+    std::optional<GtkTreeIter> m_iter;
+};
+
+// Custom container class
+template <typename ModelType>
+class GtkTreeModelContainer
+{
+public:
+    using TreeModelIterator = GtkTreeModelIterator<ModelType>;
+
+    GtkTreeModelContainer(GtkTreeModel* model)
+        : m_model{model}
+        , m_begin_iter{TreeModelIterator(m_model, std::nullopt)}
+        , m_end_iter{TreeModelIterator(m_model, std::nullopt)}
+    {
+        GtkTreeIter iter;
+        gtk_tree_model_get_iter_first(m_model, &iter);
+        m_begin_iter = TreeModelIterator(m_model, iter);
+    }
```

If this last commit is desirable, this ctor will init m_begin_iter with a dummy value before finding the proper beginning iterator. If we remove the initializer line then compilation fails.